File: app/src/main/java/com/chrispoole/intervaltimer/model/TimerUiState.kt
package com.chrispoole.intervaltimer.model

/** Immutable snapshot the service publishes and the UI collects. The single contract. */
data class TimerUiState(
    val running: Boolean = false,
    val paused: Boolean = false,
    val phase: Phase = Phase.PREPARE,
    val remainingMs: Long = 0L,
    val intervalDurationMs: Long = 0L,
    val fraction: Float = 0f,
    val round: Int = 0,
    val totalRounds: Int = 0,
    /** Work sets in one pass, 0 when the workout runs once — the pips' row width. See [Workout]. */
    val roundsPerPass: Int = 0,
    val done: Boolean = false,
    /** Current section's name, "" when unnamed — rides under the phase label. */
    val label: String = "",
) {

    companion object {
        val Idle = TimerUiState()
    }
}

/** A settable duration: "45s" under a minute, "1:30" at a minute or more. */
fun secLabel(sec: Int): String =
    if (sec < 60) "${sec}s" else "${sec / 60}:${(sec % 60).toString().padStart(2, '0')}"

/**
 * Ceil-to-second clock so a fresh interval reads its full duration and only hits 0 at true zero.
 * Under a minute: bare seconds (16, 30). A minute or more: M:SS with no leading zero (1:20).
 */
fun formatMs(ms: Long): String {
    val totalSec = ((ms.coerceAtLeast(0L) + 999) / 1000).toInt()
    val min = totalSec / 60
    val sec = totalSec % 60
    return if (min == 0) sec.toString() else "$min:${sec.toString().padStart(2, '0')}"
}

File: ios/IntervalTimerCore/Sources/IntervalTimerCore/TimerUiState.swift
import Foundation

/// Immutable snapshot the engine publishes and the UI renders. The single contract.
public struct TimerUiState: Equatable, Sendable {
    public var running: Bool
    public var paused: Bool
    public var phase: Phase
    public var remainingMs: Int
    public var intervalDurationMs: Int
    public var fraction: Double
    public var round: Int
    public var totalRounds: Int
    public var done: Bool

    public init(
        running: Bool = false,
        paused: Bool = false,
        phase: Phase = .prepare,
        remainingMs: Int = 0,
        intervalDurationMs: Int = 0,
        fraction: Double = 0,
        round: Int = 0,
        totalRounds: Int = 0,
        done: Bool = false
    ) {
        self.running = running
        self.paused = paused
        self.phase = phase
        self.remainingMs = remainingMs
        self.intervalDurationMs = intervalDurationMs
        self.fraction = fraction
        self.round = round
        self.totalRounds = totalRounds
        self.done = done
    }

    public static let idle = TimerUiState()
}

/// A settable duration: "45s" under a minute, "1:30" at a minute or more.
public func secLabel(_ sec: Int) -> String {
    sec < 60 ? "\(sec)s" : "\(sec / 60):\(pad2(sec % 60))"
}

/// Ceil-to-second clock so a fresh interval reads its full duration and only hits 0 at true zero.
/// Under a minute: bare seconds (16, 30). A minute or more: M:SS with no leading zero (1:20).
public func formatMs(_ ms: Int) -> String {
    let totalSec = (max(ms, 0) + 999) / 1000
    let min = totalSec / 60
    let sec = totalSec % 60
    return min == 0 ? "\(sec)" : "\(min):\(pad2(sec))"
}

private func pad2(_ n: Int) -> String { n < 10 ? "0\(n)" : "\(n)" }
